### ISSUE TITLE
Encountering problems running locally

### ISSUE THREAD

**Issue by GoodManWEN** [dabeaz/wadze#3]
Hi dabeaz!

Big fan of you since your live demo at PyCon Brasil 2015. 

I've just discovered your keynote video at PyCon India, 2019, I felt very impressed and want to try the python-wasm interpreter myself. Following your hint I found the original repository of rust-written rocket games, fortunately I've learned a little bit about rust so I was able to get it compiled and run smoothly in a javascript environment.

Things started to go wrong when I tried to interpret it using python. I followed your instructions in the video step by step to restore a stack virtual machine, but wadze seems to have a problem that it translates the wasm file into VM code that contains some wierd command that you didn't mention in live demo such as `('global.get', 0)` & `('global.set', 0)`, I checked the commit history of the rust rocket game and found that it has not changed since 2019(which means that the results of my compilation should not make a huge difference), I'm not sure if this unknown command comes due to the upgrade of the rust compiler, or due to I'm compiling on a windows platform. 

Then I tried to solve this problem. Although my understanding of VM is not yet thorough, but anyway I thought that since they have only one location for the operation of `globals`, it might be fine to create a new `globals` dictionary and bind it directly to the virtual machine itself and imitate the behavior of `locals` to operate on it.

However, during testing the code that calls window initialization, I found that according to the byte code interpretation, `globals` seem to require some default value and that value will participate in the following operations. Since I have set an initial value of `0.0` for `globals`, it seems to cause errors in subsequent code runs.

```Python
# Python trigger at rocket.py

# Call something
width = float64(800.0)
height = float64(600.0)
m.call(exports['resize'], width, height)   # Prayer
```

```Python
# Translated wasm byte codes of the first line of exports['resize']

Code(
    locals=['i32', 'i32', 'i64'], 
    instructions=[
        ('global.get', 0),       # The error line 
        ('i32.const', 32), 
        ('i32.sub',), 
        ('local.tee', 3), 
        ('global.set', 0),       # The error line 
        ('block', None, [('local.get', 1), ('local.get', 2), ('i32.add',), ('local.tee', 2), ('local.get', 1), ('i32.lt_u',), ('br_if', 0), ('local.get', 0), ('i32.const', 4), ('i32.add',), ('i32.load', 2, 0), ('local.tee', 1), ('i32.const', 1), ('i32.shl',), ('local.tee', 4), ('local.get', 2), ('local.get', 4), ('local.get', 2), ('i32.gt_u',), ('select',), ('local.tee', 2), ('i32.const', 4), ('local.get', 2), ('i32.const', 4), ('i32.gt_u',), ('select',), ('i64.extend_i32_u',), ('i64.const', 24), ('i64.mul',), ('local.tee', 5), ('i64.const', 32), ('i64.shr_u',), ('i32.wrap_i64',), ('i32.eqz',), ('i32.const', 3), ('i32.shl',), ('local.set', 2), ('local.get', 5), ('i32.wrap_i64',), ('local.set', 4), ('block', None, [('block', None, [('local.get', 1), ('i32.eqz',), ('br_if', 0), ('local.get', 3), ('i32.const', 16), ('i32.add',), ('i32.const', 8), ('i32.add',), ('i32.const', 8), ('i32.store', 2, 0), ('local.get', 3), ('local.get', 1), ('i32.const', 24), ('i32.mul',), ('i32.store', 2, 20), ('local.get', 3), ('local.get', 0), ('i32.load', 2, 0), ('i32.store', 2, 16), ('br', 1)]), ('local.get', 3), ('i32.const', 0), ('i32.store', 2, 16)]), ('local.get', 3), ('local.get', 4), ('local.get', 2), ('local.get', 3), ('i32.const', 16), ('i32.add',), ('call', 6), ('block', None, [('local.get', 3), ('i32.load', 2, 0), ('i32.const', 1), ('i32.ne',), ('br_if', 0), ('local.get', 3), ('i32.const', 8), ('i32.add',), ('i32.load', 2, 0), ('local.tee', 0), ('i32.eqz',), ('br_if', 1), ('local.get', 3), ('i32.load', 2, 4), ('local.get', 0), ('call', 142), ('unreachable',)]), ('local.get', 3), ('i32.load', 2, 4), ('local.set', 1), ('local.get', 0), ('i32.const', 4), ('i32.add',), ('local.get', 3), ('i32.const', 8), ('i32.add',), ('i32.load', 2, 0), ('i32.const', 24), ('i32.div_u',), ('i32.store', 2, 0), ('local.get', 0), ('local.get', 1), ('i32.store', 2, 0), ('local.get', 3), ('i32.const', 32), ('i32.add',), ('global.set', 0), ('return',)]), 
        ('call', 144), 
        ('unreachable',)
    ]
)
```

```
# Error raised by cpython interperter

Traceback (most recent call last):
  File "C:\Users\Wen\Documents\python-rocket-wasm\rocket.py", line 82, in <module>
    m.call(exports['resize'], width, height)   # Prayer
  File "C:\Users\Wen\Documents\python-rocket-wasm\machine.py", line 164, in call
    self.execute(func.code, locals)
  File "C:\Users\Wen\Documents\python-rocket-wasm\machine.py", line 244, in execute
    self.execute(args[1], locals)
  File "C:\Users\Wen\Documents\python-rocket-wasm\machine.py", line 199, in execute
    self.push(locals[args[0]])
KeyError: 2

* where locals is {0: 800.0, 1: 600.0, 3: -32}
```

Since I am only an enthusiastic audience but not an expert of wasm, nor VM. At this point I think it's beyond my ability to solve the problem, I can't understand what's going wrong not to mention to fix it. Think I should raise an issue here and wait if you can give any help to let me run the super cool live demo on my own machine.

I've committed the complete code (that I copied from the video), as well as the compiled wasm file, to [https://github.com/GoodManWEN/python-rocket-wasm](https://github.com/GoodManWEN/python-rocket-wasm), not sure if this helps.

Thanks anyway.

**Comment by dabeaz** [comment]
It's really hard for me to say exactly.  I never compiled the Rocket game from Rust, but used the binary provided on the website directly.   The following gist has code from the talk  https://gist.github.com/dabeaz/7d8838b54dba5006c58a40fc28da9d5a

In principle, the global.set/get opcode can manipulate a dictionary.   However, you'd probably need to carefully read the WASM spec get the meaning.   There may be some initialization code that needs to execute as well.

**Comment by GoodManWEN** [comment]
Thanks for reply! Later I wrote a small program to determine if all of the operation that translated byte code contains were in those VM could understand, and it seems that the rust compiler has changed a lot of over the years and there're several unrecognized commands occurring, such as `global.get`,`global.set`,`call_indirect`,`f64.abs`,`f64.neg`,`f64.floor`,`f64.copysign`etc.

Considering that there are these many features that I need to implement myself, I prefer the option of downloading the compiled version as you said, I was able to run my program successfully and get the correct results, It is as cool as imagined.

During the first few runs, the program prompt that it cannot find operation `call_indirect`, but this error magically disappeared in subsequent runs.
